### PR TITLE
Add access restrictions for Discord Signup

### PR DESCRIPTION
**Pull request by rshipp**
### Description

Deny access to Discord Signup to:
1. unenrolled people (from the policy)
2. people with a TA role that does NOT have a TA volunteer role (e.g. staff, faculty, or test volunteer roles) (from the controller, because doing it from the policy would be a pain)

### Motivation and Context

1. We don't want people who aren't enrolled in a course to be added to the discord server for that course
2. We don't want Staff to be added to the discord server through this page
3. We don't want anyone with some kind of "guest" enrollment to be added to any discord server

Task: https://app.asana.com/0/0/1201314155101899/f

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have added neccessary labels
- [ ] My PR title follows the convention in the guide
- [ ] I have assigned myself to the PR
- [ ] I have requested reviews from team members
- [ ] I have created a task for reviewers on PM tool
- [ ] All new and existing tests passed.

### CODE REVIEW

File: app/controllers/discord_signups_controller.rb [hunk 1]
```diff
@@ -26,7 +26,20 @@ class DiscordServerIdError  < StandardError; end
...
+    course = Course.find_by_canvas_course_id!(@lti_launch.course_id)
+    authorize course, policy_class: DiscordSignupPolicy
+
+    program_id = course.salesforce_program_id
+
+    participant = SalesforceAPI.client.find_participant(
+      contact_id: current_user.salesforce_id, program_id: program_id
+    )
+
+    # Show Staff/Faculty/Test TAs a different page, so they can't use this.
```
Comment (sadleb) [inline app/controllers/discord_signups_controller.rb:39]: There is a `SalesforceAPI.is_teaching_assistant?` method that treats staff and other TA roles as a Teaching Assistant. This is new “is actual TA?” logic. I’m not sure what to do in this PR, but I’d rather have all the methods that determine your “role” based on the participant be consolidated there for maintainability. Thoughts?

Comment (sadleb) [reply app/controllers/discord_signups_controller.rb:39]: Maybe just put this in a `is_real_teaching_assistant?` method over there for now so at least it’s all in the same spot and we’re more likely to catch this particular use case as we continue to refactor / refine?

Comment (rshipp) [reply app/controllers/discord_signups_controller.rb:39]: I had it over there originally, but there were so many moving pieces / discussion that i pulled it back out. I can put it back in a new method, that seems fine for now.

**Review by sadleb — APPROVED** [review]
Lgtm

File: app/controllers/discord_signups_controller.rb [hunk 2]
```diff
@@ -26,7 +26,20 @@ class DiscordServerIdError  < StandardError; end
...
+    if SalesforceAPI.is_real_teaching_assistant?(participant)
+      return render :no_discord
+    end
```
Comment (elenabaurkot) [inline app/controllers/discord_signups_controller.rb:43]: I'm not sure if I have a full understanding of this part. In the description it says to deny access to people with a TA role that does NOT have a TA volunteer role. From what I can understand, this method returns true for people who have a TA role and a TA volunteer role. So would this prevent the actual TAs from accessing the page? @rshipp

Comment (rshipp) [reply app/controllers/discord_signups_controller.rb:43]: yep this is backwards, i'll revert my earlier commit

**Review by elenabaurkot — COMMENTED** [review]
lgtm 👍

File: app/controllers/discord_signups_controller.rb (hunk 1 above)
Comment (rshipp) [reply app/controllers/discord_signups_controller.rb:39]: https://github.com/bebraven/platform/pull/916#discussion_r752285727

**Pull request merged**